Plutarch/FFI.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.FFI (
+  foreignExport,
+  foreignImport,
+  unsafeForeignExport,
+  unsafeForeignImport,
+) where
+
+import Data.Kind (Constraint, Type)
+import Data.Text (Text)
+import GHC.TypeLits (TypeError)
+import qualified GHC.TypeLits as TypeLits
+import qualified Generics.SOP as SOP
+import Plutarch.Bool (PBool)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal (
+  ClosedTerm,
+  PDelayed,
+  PType,
+  RawTerm (RCompiled),
+  Term (Term),
+  TermResult (TermResult),
+  asClosedRawTerm,
+  compile',
+  (:-->),
+ )
+import Plutarch.Internal.PlutusType (PlutusType (PInner))
+import Plutarch.String (PString)
+import Plutus.V1.Ledger.Scripts (Script (unScript), fromCompiledCode)
+import PlutusTx.Code (CompiledCode, CompiledCodeIn (DeserializedCode))
+import PlutusTx.Prelude (BuiltinString)
+import UntypedPlutusCore (fakeNameDeBruijn)
+import qualified UntypedPlutusCore as UPLC
+
+data ForallPhantom :: Type
+data PhorallPhantom :: PType
+
+foreignExport :: PlutarchInner p PhorallPhantom ~~ PlutusTxInner t ForallPhantom => ClosedTerm p -> CompiledCode t
+foreignExport = unsafeForeignExport
+
+foreignImport :: PlutarchInner p PhorallPhantom ~~ PlutusTxInner t ForallPhantom => CompiledCode t -> ClosedTerm p
+foreignImport = unsafeForeignImport
+
+unsafeForeignExport :: ClosedTerm p -> CompiledCode t
+unsafeForeignExport t = DeserializedCode program Nothing mempty
+  where
+    program =
+      UPLC.Program () (UPLC.Version () 1 0 0) $
+        UPLC.termMapNames fakeNameDeBruijn $
+          compile' $
+            asClosedRawTerm t
+
+unsafeForeignImport :: CompiledCode t -> ClosedTerm p
+unsafeForeignImport c = Term $ const $ TermResult (RCompiled $ UPLC.toTerm $ unScript $ fromCompiledCode c) []
```

Comment:
Rather than returnining `p`, you should return `POpaque`, and it wouldn't be unsafe either.

Reply:
Can you explain your motivation? `POpaque` was literally made for this.

Reply:
I've added `PUnit` which can provide the same functionality.

Reply:
I can't see how `PUnit` covers this? `PUnit` has only one legal term, `λx.x`. Any term is a legal `POpaque`.

Reply:
I meant it covers the case _when you don't care about the type and only the fact that it's evaluated_.

Reply:
That isn't true from a performance perspective. You need to change the term to make it of type `()`.

Reply:
I tried adding `data Opaque = forall a. Opaque a` to mirror `POpaque` on the Haskell side, but `plutus-tx-plugin` can't handle the constructor:
```
GHC Core to PLC plugin: E043:Error: Reference to a name which is not a local, a builtin, or an external INLINABLE function: Type variable: a
Context: Compiling type: a
Context: Compiling data constructor type: Plutarch.FFI.Opaque
```

Reply:
Because I don't see how `CompiledCode a -> ClosedTerm POpaque` improves things. I already have working bijections for most common types, including user-defined ones.

I can add `CompiledCode Any` <-> `ClosedTerm POpaque`. What that be enough?

Reply:
I think I finally understand what you meant. Do I?

Reply:
See the last commit.